src/service/workspace_wrapper.py

```diff
@@ -20,23 +23,57 @@ def __init__(self, workspace: Workspace):
         """
         self._ws = workspace
 
+    def _get_type(self, obj_info) -> str:
+        return obj_info[2].split('-')[0]
+
     def get_object_metadata(
         self,
         upas: list[str],
-        allowed_types: set[str] | None = None
+        allowed_types: set[str] | None = None,
+        allowed_set_types: set[str] | None = None,
     ) -> list[dict[str, Any]]:
-        """
+        f"""
         Get metadata for selected workspace objects.
 
         upas - the UPAs of objects from which to retrieve metadata.
         allowed_types - the allowed types of the objects. If object types are not in this list,
             an error will be thrown.
+        allowed_set_types - allowed types of object sets. The object sets are expanded by assuming
+            all their references are to target objects (e.g. reference context in the object is
+            not taken info account), and those objects are checked against allowed_types if
+            provided.
 
-        Returns a list of the metadata in the same order as the UPAs.
+        Returns a list of the metadata, adding a special field, {WORKSPACE_UPA_PATH},
+        to the metadata, with the path to the workspace object in string format, e.g. 3/2/1;6/5/4.
         """
         # The workspace error format really sucks. Error codes are bascially a necessity for
         # handling errors reasonably - trying to figure out the error type from an arbitrary
         # string is not reasonable
+        std_objs, set_objs = self._get_object_info(upas, allowed_types, allowed_set_types)
+        std_objs2 = []
+        if set_objs:
+            paths = [meta[WORKSPACE_UPA_PATH] for meta in set_objs]
+            res = self._ws.get_objects2(
+                {"objects": [{'ref': p} for p in paths], "no_data": 1, 'ignoreErrors': 1}
+            )['data']
+            set_upas = set()
+            for path, o in zip(paths, res):
+                if not o:
+                    # Race condition - the workspace allowed get_info3 but denied on get_objects2
+                    # Need to test with mocked workspace client
+                    raise errors.DataPermissionError(
+                        f"The workspace service disallowed acces to object {path}"
+                    )
+                set_upas.update([f'{path};{r}' for r in o['refs']])
+            if len(set_upas) > 10000:
```

Comment:
using `match_retrieval.MAX_UPAS` here?

Reply:
This limit is slightly different than max upas as it's there because I'm lazy and didn't want to write a loop over get_object_info3 to do more than 10000 set objects. It might go away in the future, or the limit in match_retrieval might go away and this stays the same, but either way with the current implementation can't handle more than 10000 set UPAs even if we got rid of this check. I just put it here so we'd know if this happened, which could give us the impetus to write the loop or make a concrete decision to not support more than X UPAs in a match. As such my preference is to leave it alone for now

Reply:
👍